src/common/src/array/list_array.rs

```diff
@@ -431,118 +376,189 @@ impl ListValue {
+
+impl<T: PrimitiveArrayItemType> FromIterator<T> for ListValue {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<PrimitiveArray<T>>().into())
+    }
+}
+
+impl FromIterator<bool> for ListValue {
+    fn from_iter<I: IntoIterator<Item = bool>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<BoolArray>().into())
+    }
+}
+
+impl<'a> FromIterator<Option<&'a str>> for ListValue {
+    fn from_iter<I: IntoIterator<Item = Option<&'a str>>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<Utf8Array>().into())
+    }
+}
+
+impl<'a> FromIterator<&'a str> for ListValue {
+    fn from_iter<I: IntoIterator<Item = &'a str>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<Utf8Array>().into())
+    }
+}
+
+impl FromIterator<ListValue> for ListValue {
+    fn from_iter<I: IntoIterator<Item = ListValue>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<ListArray>().into())
+    }
+}
+
+impl From<ListValue> for ArrayImpl {
+    fn from(value: ListValue) -> Self {
+        *value.values
     }
 }
 
 #[derive(Copy, Clone)]
-pub enum ListRef<'a> {
-    Indexed { arr: &'a ListArray, idx: usize },
-    ValueRef { val: &'a ListValue },
+pub struct ListRef<'a> {
+    array: &'a ArrayImpl,
+    start: u32,
+    end: u32,
 }
 
 impl<'a> ListRef<'a> {
     /// Returns the length of the list.
     pub fn len(&self) -> usize {
-        match self {
-            ListRef::Indexed { arr, idx } => (arr.offsets[*idx + 1] - arr.offsets[*idx]) as usize,
-            ListRef::ValueRef { val } => val.values.len(),
-        }
+        (self.end - self.start) as usize
     }
 
     /// Returns `true` if the list has a length of 0.
     pub fn is_empty(&self) -> bool {
-        self.len() == 0
+        self.start == self.end
     }
 
-    /// Returns the elements in the flattened list.
-    pub fn flatten(self) -> Vec<DatumRef<'a>> {
-        // XXX: avoid using vector
-        iter_elems_ref!(self, it, {
-            it.flat_map(|datum_ref| {
-                if let Some(ScalarRefImpl::List(list_ref)) = datum_ref {
-                    list_ref.flatten()
-                } else {
-                    vec![datum_ref]
-                }
-                .into_iter()
-            })
-            .collect()
-        })
+    /// Returns the data type of the elements in the list.
+    pub fn data_type(&self) -> DataType {
+        self.array.data_type()
     }
 
-    /// Returns the total number of elements in the flattened list.
-    pub fn flatten_len(self) -> usize {
-        iter_elems_ref!(self, it, {
-            it.map(|datum_ref| {
-                if let Some(ScalarRefImpl::List(list_ref)) = datum_ref {
-                    list_ref.flatten_len()
-                } else {
-                    1
-                }
-            })
-            .sum()
-        })
+    /// Returns the elements in the flattened list.
+    pub fn flatten(self) -> ListRef<'a> {
```

Comment:
I realize that there's a subtle change on the behavior.

`array[array[1, 2], null, array[3, null]]`

- Previously, this was flatten into `array[1, 2, null, 3, null]`.
- After this PR, this will be flatten into `array[1, 2, 3, null]`.

The new behavior seems more accurate. However, Postgres simply rejects this:
```
ERROR:  multidimensional arrays must have array expressions with matching dimensions
```